code/hud/hudtarget.cpp

```diff
@@ -3979,46 +3970,117 @@ void HudGaugeLeadIndicator::renderLeadCurrentTarget()
 		dist_to_target = hud_find_target_distance(targetp, Player_obj);
 	}
 
-	srange = ship_get_secondary_weapon_range(Player_ship);
+	//Moved here to avoid potentially leaving wip undefined. These values will be
+	//overwritten if using non-default leadIndicatorBehaviors, this is desired.
+	bank_to_fire = hud_get_best_primary_bank(&prange);
+	wip = &Weapon_info[swp->primary_bank_weapons[bank_to_fire]];
 
-	if ( (swp->current_secondary_bank >= 0) && (swp->secondary_bank_weapons[swp->current_secondary_bank] >= 0) )
-	{
-		int bank = swp->current_secondary_bank;
-		tmp = &Weapon_info[swp->secondary_bank_weapons[bank]];
-		if ( !(tmp->is_homing()) && !(tmp->is_locked_homing() && Player->target_in_lock_cone) ) {
-			//The secondary lead indicator is handled farther below if it is a non-locking type
-			srange = -1.0f;
+	if (Lead_indicator_behavior != leadIndicatorBehavior::DEFAULT && Player_ship->flags[Ship::Ship_Flags::Primary_linked]) {
+		vec3d averaged_lead_pos;
+		vm_vec_zero(&averaged_lead_pos);
+		int average_instances = 0;
+		for (int i = 0; i < swp->num_primary_banks; i++) {
+			auto bank = i;
+			// calculate the range of the weapon, and only display the lead target indicator
+			// if the weapon can actually hit the target
+			Assert(bank >= 0 && bank < swp->num_primary_banks);
+			Assert(swp->primary_bank_weapons[bank] < weapon_info_size());
+
+			if (swp->primary_bank_weapons[bank] < 0)
+				continue;
+
+			wip = &Weapon_info[swp->primary_bank_weapons[bank]];
+
+			if (wip->wi_flags[Weapon::Info_Flags::Nolink] ||  (wip->wi_flags[Weapon::Info_Flags::Ballistic] && swp->primary_bank_ammo[bank] <= 0))
+				continue;
+
+			if (bank < 0)
+				continue;
+
+			if (pm->n_guns && bank != -1 ) {
+				rel_pos = &pm->gun_banks[bank].pnt[0];
+			} else {
+				rel_pos = NULL;
+			}
+
+			source_pos = Player_obj->pos;
+			if (rel_pos != NULL) {
+				vec3d	gun_point;
+				vm_vec_unrotate(&gun_point, rel_pos, &Player_obj->orient);
+				vm_vec_add2(&source_pos, &gun_point);
+			}
+
+			auto weapon_range = MIN((wip->max_speed * wip->lifetime), wip->weapon_range);
+			if (weapon_range < dist_to_target)
+				continue;
+
+			frame_offset = pickFrame(weapon_range, srange, dist_to_target);
+
+			if (frame_offset >= 0) {
+				hud_calculate_lead_pos(&lead_target_pos, &target_pos, targetp, wip, dist_to_target);
+				if (Lead_indicator_behavior == leadIndicatorBehavior::MULTIPLE) {
+					renderIndicator(frame_offset, targetp, &lead_target_pos);
+				}
+				else if (Lead_indicator_behavior == leadIndicatorBehavior::AVERAGE) {
+					vm_vec_add2(&averaged_lead_pos, &lead_target_pos);
+					average_instances++;
+				}
+			}
+		}
+		if (Lead_indicator_behavior == leadIndicatorBehavior::AVERAGE) {
+			averaged_lead_pos = averaged_lead_pos/i2fl(average_instances);
+			renderIndicator(frame_offset, targetp, &averaged_lead_pos);
 		}
```

Comment:
What does this do for average_instances == 0?
Should this if not also explicitly check for that, and in that case, not render any indicator?